src/stores/room-list/algorithms/list-ordering/NaturalAlgorithm.ts

```diff
@@ -21,42 +21,191 @@ import { SortAlgorithm } from "../models";
 import { sortRoomsWithAlgorithm } from "../tag-sorting";
 import { OrderingAlgorithm } from "./OrderingAlgorithm";
 import { RoomUpdateCause, TagID } from "../../models";
+import { RoomNotificationStateStore } from "../../../notifications/RoomNotificationStateStore";
+
+type NaturalCategorizedRoomMap = {
+    defaultRooms: Room[];
+    mutedRooms: Room[];
+};
 
 /**
  * Uses the natural tag sorting algorithm order to determine tag ordering. No
  * additional behavioural changes are present.
  */
 export class NaturalAlgorithm extends OrderingAlgorithm {
+    private cachedCategorizedOrderedRooms: NaturalCategorizedRoomMap = {
+        defaultRooms: [],
+        mutedRooms: [],
+    };
     public constructor(tagId: TagID, initialSortingAlgorithm: SortAlgorithm) {
         super(tagId, initialSortingAlgorithm);
     }
 
     public setRooms(rooms: Room[]): void {
-        this.cachedOrderedRooms = sortRoomsWithAlgorithm(rooms, this.tagId, this.sortingAlgorithm);
+        const { defaultRooms, mutedRooms } = this.categorizeRooms(rooms);
+
+        this.cachedCategorizedOrderedRooms = {
+            defaultRooms: sortRoomsWithAlgorithm(defaultRooms, this.tagId, this.sortingAlgorithm),
+            mutedRooms: sortRoomsWithAlgorithm(mutedRooms, this.tagId, this.sortingAlgorithm),
+        };
+        this.buildCachedOrderedRooms();
     }
 
     public handleRoomUpdate(room: Room, cause: RoomUpdateCause): boolean {
         const isSplice = cause === RoomUpdateCause.NewRoom || cause === RoomUpdateCause.RoomRemoved;
-        const isInPlace = cause === RoomUpdateCause.Timeline || cause === RoomUpdateCause.ReadReceipt;
+        const isInPlace =
+            cause === RoomUpdateCause.Timeline ||
+            cause === RoomUpdateCause.ReadReceipt ||
+            cause === RoomUpdateCause.PossibleMuteChange;
+        const isMuted = this.isMutedToBottom && this.getRoomIsMuted(room);
+
         if (!isSplice && !isInPlace) {
             throw new Error(`Unsupported update cause: ${cause}`);
         }
 
         if (cause === RoomUpdateCause.NewRoom) {
```

Comment:
There is a lot of if conditions, some additional comments are welcomed :)

Maybe split the content of the function in multiple private functions to make it easier to read.